src/main/java/org/jitsi/videobridge/util/PartitionedByteBufferPool.java

```diff
@@ -16,171 +16,362 @@
+         * Return rate in requests per second over the last 10 seconds.
+         */
+        private final RateStatistics returnRate
+            = new RateStatistics(1000, 1000);
+
+        /**
+         * Initializes a new partition.
+         * @param id
+         * @param initialSize
+         */
+        Partition(int id, int initialSize)
         {
             this.id = id;
             for (int i = 0; i < initialSize; ++i)
             {
-                pool.add(new byte[1500]);
+                pool.add(new byte[DEFAULT_BUFFER_SIZE]);
             }
-            System.out.println("Pool " + id + " initial fill done, size is " + pool.size());
         }
 
+        /**
+         * Returns a buffer from this pool (allocates a new one if necessary).
+         * @param requiredSize the minimum size.
+         *
+         * @return a {@code byte[]} with size at least {@code requiredSize}.
+         */
         private byte[] getBuffer(int requiredSize)
         {
             if (ByteBufferPool.ENABLE_BOOKKEEPING)
             {
-                System.out.println("partition " + id + " request number " + (numRequests.get() + 1) +
-                        ", pool has size " + pool.size());
+                logger.info("partition " + id + " request number "
+                        + (numRequests.get() + 1) + ", pool has size "
+                        + pool.size());
             }
-            byte[] buf;
-            int numTries = 0;
-            while (true) {
-                buf = pool.poll();
-                if (buf == null) {
-                    buf = new byte[1500];
-                    if (ByteBufferPool.ENABLE_BOOKKEEPING)
-                    {
-                        numEmptyPoolAllocations.incrementAndGet();
-                        numAllocations.incrementAndGet();
-                    }
-                    break;
-                } else if (buf.length < requiredSize) {
-                    if (ByteBufferPool.ENABLE_BOOKKEEPING)
-                    {
-                        System.out.println("Needed buffer of size " + requiredSize + ", got size " + buf.length +
-                                " retrying");
-                    }
 
-                    pool.offer(buf);
-                    numTries++;
-                    if (numTries == 5) {
-                        buf = new byte[1500];
-                        if (ByteBufferPool.ENABLE_BOOKKEEPING)
-                        {
-                            numWrongSizeAllocations.incrementAndGet();
-                            numAllocations.incrementAndGet();
-                        }
-                        break;
-                    }
-                } else {
-                    if (ByteBufferPool.ENABLE_BOOKKEEPING)
-                    {
-                        numNoAllocationNeeded.incrementAndGet();
-                    }
-                    break;
+            if (enableStatistics)
+            {
+                numRequests.incrementAndGet();
```

Comment:
so here we only increment numRequests if stats is enabled, but at the top level i think we always do.  is that intentional?